jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java

```diff
@@ -719,4 +727,23 @@ public static String buildPersistenceUnitName(URL url, String puName){
        return fullPuName;
    }
 
+    /**
+     * @param file archive file URL
+     * @param entry a directory entry in the archive (or an empty string)
+     * @return true if the file-entry pair can be a persistence root according
+     * to JPA Spec (8.2).
+     */
+    private static boolean isValidRootInArchive(String file, String rootEntry) {
```

Comment:
I feel like we should just put this logic in the ```computePURootURL``` method. There we already have the protocol so ```extension``` is available, this helper method is only called once anyway, and we can throw an Exception instead of bothering with boolean checks

Reply:
> I feel like we should just put this logic in the computePURootURL

In my opinion, `computePURootURL` is getting unwieldy as it is. This validation doesn't interact with the computation algorithm itself, so for me it makes sense to name and separate this part of code. This way I can keep the computation simpler as well - just cut off the descriptor, clean up and we're done.

> There we already have the protocol so extension is available,

Protocol here doesn't determine the extension. We have three protocols (`jar:`, `wsjar:`, `zip:`), and two extensions of interest (`.jar`, `.war`), and all six combinations are possible (as far as I see).

Reply:
Ok, you convinced me :)